[FEAT] Add streaming + parallel CSV reader, with decompression support.

clarkzinzow opened the pull request:
This PR adds streaming + parallel CSV reading and parsing, along with support for streaming decompression. In particular, this PR:
- Adds support for streaming decompression for brotli, bz, deflate, gzip, lzma, xz, zlib, and zstd.
- Performs chunk-based streaming CSV reads, filling up a small buffer of unparsed records.
- Pipelines chunk-based CSV parsing with reading by spawning Tokio + rayon parsing tasks.
- Performances chunk parsing, as well as column parsing within a chunk, in parallel on the rayon threadpool.
- Changes schema inference to involve an (at most) 1 MiB file peak rather than a full file read.
- Gathers a mean row size in bytes estimate during schema inference and propagates this estimate back to the reader.
- Unifies local and cloud reads + schema inference.
- Adds thorough Rust-side local + cloud test coverage.

The streaming + parallel reading + parsing leads to a 4-8x speed up over the pyarrow reader and the previous non-parallel reader when benchmarking large file (~1 GB) reads, while also resulting in lower memory utilization due to the streaming reading + parsing.

## TODOs (follow-up PRs)

- [ ] Add snappy decompression support (need to essentially do something like [this](https://github.com/belltoy/tokio-snappy/blob/master/src/lib.rs))

samster25 commented on src/daft-csv/src/read.rs:
```diff
@@ -54,57 +77,159 @@ async fn read_csv_single(
+                schema,
+                // Default buffer size of 512 KiB.
+                buffer_size.unwrap_or(512 * 1024),
+                // Default chunk size of 64 KiB.
```
Do these constants make sense given its a local buffered file?

clarkzinzow replied:
I have a TODO to benchmark locally and tweak these, but I'm assuming that tweaking these won't matter as much for local reads as they do for cloud reads. I can look into tweaking these if you'd like!

samster25 commented on src/daft-csv/src/read.rs:
```diff
@@ -54,57 +77,159 @@ async fn read_csv_single(
+                // Default chunk size of 64 KiB.
+                chunk_size.unwrap_or(64 * 1024),
+                // Default max chunks in flight would result in 2 * 8 * 1024 * chunk_size bytes, or 1 GiB with the default chunk size.
+                max_chunks_in_flight.unwrap_or(8 * 1024),
```
Why is this number so high given that the parallelism is limited by the number rayon tasks that can run in parallel which would be limited by the number of cores on a machine. Then the max number in flight could just be the `num_cpus` on the current machine?

clarkzinzow replied:
Opted for 2x # of cores so we can (1) pipeline I/O-bound reading with compute-bound parsing, and (2) try to ensure that each worker thread will have a downloaded chunk immediately available for parsing even if parsing is completing in ~uniform waves.

samster25 reviewed (commented):
Looks good! And you were right haha, I think we should probably pull out the estimated rows commit until after we do the MicroPartition and scan operator work.

clarkzinzow commented on daft/table/schema_inference.py:
```diff
@@ -60,13 +60,13 @@ def from_csv(
     file: FileInput,
     storage_config: StorageConfig | None = None,
-) -> Schema:
+) -> tuple[Schema, int | None]:
```
I made it opt-in for others.

clarkzinzow commented on src/daft-csv/src/metadata.rs:
```diff
@@ -1,68 +1,250 @@
+        let first_record_count = record.len();
+        (
+            (0..first_record_count)
+                .map(|i| format!("column_{}", i + 1))
```
Initially tried to make this a const, but `format!()` strings need to be string literals, and I didn't want to pull in a crate like `const_format` just for this.

clarkzinzow commented on src/daft-csv/src/metadata.rs:
```diff
@@ -1,68 +1,250 @@
+                // default to Utf8 for conflicting datatypes (e.g bool and int)
+                DataType::Utf8
+            }
+        }
```
I'm assuming that we'll have a variety of follow-ups there, yes.

clarkzinzow commented:
@samster25 I reverted the estimated row size piping from schema inference, added a good bit more test coverage, and addressed your primary review comments, PTAL!

samster25 commented on src/daft-csv/src/read.rs:
```diff
@@ -190,119 +289,1081 @@ where
+        let mut rows_read = 1;
+        // Total number of rows read across all reads.
+        let mut total_rows_read = 0;
+        while rows_read > 0 && total_rows_read < num_rows {
```
What about computing `estimated_mean_row_size` as a running average of `bytes_read / rows_read `

clarkzinzow replied:
Working on that now!

samster25 commented on src/daft-csv/src/read.rs:
```diff
@@ -54,57 +76,177 @@ async fn read_csv_single(
+    schema: Option<SchemaRef>,
+    buffer_size: Option<usize>,
+    chunk_size: Option<usize>,
+    max_chunks_in_flight: Option<usize>,
```
do we still need `max_chunks_in_flight` threaded all the way to the top now that we default to num cores?

samster25 commented on daft/runners/partitioning.py:
```diff
@@ -44,10 +44,16 @@ class TableParseCSVOptions:
         header_index: Index of the header row, or None if no header
+        buffer_size: Size of the buffer (in bytes) used by the streaming reader.
+        chunk_size: Size of the chunks (in bytes) deserialized in parallel by the streaming reader.
+        max_chunks_in_flight: Maximum number of chunks deserialized concurrently by the streaming reader.
```
I dont think we need to expose `max_chunks_in_flight` to the user

clarkzinzow replied:
I was thinking that it would be nice to expose an override to the user in case our default is over-throttling, so the user doesn't have to wait for another release in order to see a performance fix, but we can hide this arg if we consider that to be unlikely given our concurrency/parallelism model.

samster25 reviewed (approved):
Looks good! However we should also test everything through the python side as well via Dataframe tests if that isn't already done!
